### PR TITLE
roles: add Roles

### PR DESCRIPTION
**Pull request by jac**
Adds Roles support to the SDK.
Moves Services into a separate package so that it can be imported by both Roles and Scopes.

Closes CORE-204
## Test plan
CI

### CODE REVIEW

File: roles/roles.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
+package roles
+
+import (
+	"slices"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/services"
+)
+
+// Role is the string literal of a role.
+type Role string
```
Comment (bobheadxi) [inline roles/roles.go:10]: I think we might want to mirror the design of the `scopes` package, where `type Role string` is a fully qualified name (`service::role`)

Comment (bobheadxi) [reply roles/roles.go:10]: Judging from tests and the structure of `RegisteredRoles`, it looks like right now `Role` is the second part of a fully qualified name (only `role`)

File: roles/roles.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+// RoleRegistration is a struct that contains the display name and roles for a service.
+type RoleRegistration struct {
```
Comment (bobheadxi) [inline roles/roles.go:13]: Should we just call this 
```suggestion
type EntitleMetadata struct {
```

for clarity? the display name shouldn't be used elsewhere

File: roles/roles.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	DisplayName string
+	Roles       []Role
+}
+
+var (
+	// services.Dotcom
+	dotcomDisplayName string = "Sourcegraph Dotcom"
```
Comment (bobheadxi) [inline roles/roles.go:20]: this probably doesn't need to be a variable, it can just be set at the callsite

File: roles/roles.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	// Roles for services.Dotcom
+
+	// RoleDotcomSiteadmin is the role for site admins on dotcom
+	RoleDotcomSiteadmin Role = "site_admin"
+)
+
+// Registered returns a map of all registered roles for all services.
+type RegisteredRoles map[services.Service]RoleRegistration
+
+func Registered() RegisteredRoles {
```
Comment (bobheadxi) [inline roles/roles.go:36]: IMO we should have 2 explicit things:

1. A helper that surfaces the Entitle version of the roles set (`[]EntitleRoleRegistration`)
2. A helper that surfaces `[]Role` with helpers on `type Roles []Role` like `(Roles).Contains(service, role)` or similar

File: roles/roles.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	"slices"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/services"
+)
+
+// Role is always the full qualified role name.
+type Role string
+
+// ToRole returns a role string in the format of
+// "service::role".
```
Comment (bobheadxi) [inline roles/roles.go:13]:
```suggestion
// "service::rolename", which comprises a fully qualified role name.
```

File: roles/roles.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+// ToStrings converts a list of roles to a list of strings.
+func ToStrings(roles []Role) []string {
+	ss := make([]string, len(roles))
+	for i, role := range roles {
+		ss[i] = string(role)
+	}
+	return ss
+}
+
+// ToRoles converts a list of strings to a list of roles.
```
Comment (bobheadxi) [inline roles/roles.go:27]:
```suggestion
// ToRoles converts a list of strings to a list of roles.
// It does not validate each input value.
```

File: roles/roles.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	dotcomRoles = []string{
+		RoleNameDotcomSiteAdmin,
+	}
+)
+
+const (
+	// Role names for services.Dotcom
+
+	// Dotcom site admin
+	RoleNameDotcomSiteAdmin = "site_admin"
```
Comment (bobheadxi) [inline roles/roles.go:47]: We should only ever export fully qualified names:

```suggestion
	RoleDotComSiteAdmin = ToRole(service.DotCom, "site_admin")
```

Comment (jac) [reply roles/roles.go]: What do you think of https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/57/commits/dc28f5314ed9de439978dfa93d7c64aa2fa45f3a ?

In this case it probably makes sense that `ToRole` isn't exported

File: roles/roles.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
+package roles
+
+import (
+	"slices"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/services"
+)
+
+// Role is always the full qualified role name.
```
Comment (unknwon) [inline roles/roles.go:9]:
```suggestion
// Role is always the full qualified role name, e.g. "dotcom::site_admin".
```

**Review by unknwon — APPROVED** [review]
Another than [Robert's comment](https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/57#discussion_r1717644177).

File: roles/roles.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	for i, s := range strings {
+		roles[i] = Role(s)
+	}
+	return roles
+}
+
+type ParsedRole struct {
+	Service services.Service
+	Name    string
+}
```
Comment (bobheadxi) [inline roles/roles.go:41]: let's add a docstrings on each, and add a `(ParsedRole).ToRole() Role` that shows what the components mean

**Review by bobheadxi — APPROVED** [review]
overall LGTM, added a suggestion!

**Pull request merged**